vyper/semantics/validation/local.py

```diff
@@ -493,50 +499,53 @@ class _LocalExpressionVisitor(VyperNodeVisitorBase):
     ignored_types = (vy_ast.Constant, vy_ast.Name)
     scope_name = "function"
 
-    def visit_Attribute(self, node: vy_ast.Attribute) -> None:
+    def visit(self, node, type_=None):
+        super().visit(node, type_)
+
+    def visit_Attribute(self, node: vy_ast.Attribute, type_: Optional[BaseTypeDefinition]) -> None:
         self.visit(node.value)
         _validate_msg_data_attribute(node)
         _validate_address_code_attribute(node)
 
-    def visit_BinOp(self, node: vy_ast.BinOp) -> None:
-        self.visit(node.left)
-        self.visit(node.right)
+    def visit_BinOp(self, node: vy_ast.BinOp, type_: Optional[BaseTypeDefinition]) -> None:
```

Comment:
we could add an alias earlier in this module, like `OptionalType = Optional[BaseTypeDefinition]`, to shorten these signatures a bit.